### PR TITLE
Test for Bayesian Optimization Algo

### PR DESCRIPTION
**Pull request by jdplatt**
#230 

Here is a PR containing tests for the code in pkg/suggestion/bayesianoptimization. You can run them using test/scripts/python_tests.sh. I have close to full coverage but didn't cover the parts of the code that deal with the beginning of tuning when there is no data. I think there is a cleaner way to handle this using a different suggestion algorithm such as random search. I fixed a few bugs in the way the models were trained alongside writing the tests.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/katib/406)

### CODE REVIEW

**Comment by YujiOshima** [comment]
@jdplatt Thank you for PR! Could you sign the CLA?

**Comment by jdplatt** [comment]
I signed it!

**Comment by jdplatt** [comment]
@YujiOshima 

Had a few issues getting corporate CLA to work but should be all good now.

**Comment by YujiOshima** [comment]
Thanks, I will take a look!

File: test/scripts/python_tests.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+pip install -r cmd/suggestion/bayesianoptimization/requirements.txt
```
Comment (YujiOshima) [inline test/scripts/python_tests.sh]: You should add this script to https://github.com/kubeflow/katib/blob/master/test/scripts/unit-test.sh or add a new workflow for python test to argo.

File: pkg/suggestion/bayesianoptimization/src/algorithm_manager.py [hunk 2]
```diff
@@ -1,39 +1,31 @@
...
     diff = np.absolute(diff)
     return feasible_values[np.argmin(diff)]
 
+
 def deal_with_categorical(feasible_values, one_hot_values):
     """ function to do the one hot encoding of the categorical values """
-    #index = np.argmax(one_hot_values)
-    index = one_hot_values.argmax()
+    index = np.argmax(one_hot_values)
+    #index = one_hot_values.argmax()
```
Comment (YujiOshima) [inline pkg/suggestion/bayesianoptimization/src/algorithm_manager.py]: Why did you change this?

Comment (jdplatt) [reply pkg/suggestion/bayesianoptimization/src/algorithm_manager.py]: The two lines perform the same task. However, the np.argmax function works on any array_like object (e.g. list, tuple, or ndarray), while the ndarray.argmax() method only works on ndarrays. I switched for two reasons, i) the old version raised an exception because it was being passed a list at some point, ii) the function approach is more flexible and requires less type casts to force array_like variables into ndarrays.

File: test/scripts/python_tests.sh (hunk 1 above)
Comment (jdplatt) [reply test/scripts/python_tests.sh]: Thanks for catching this! I added it at the end of unit-test.sh

**Comment by jdplatt** [comment]
Thanks for looking at my PR. I fixed the two issues you raised. Do you need me to do anything else?

**Comment by YujiOshima** [comment]
@jdplatt Thanks!
/ok-to-test

File: pkg/suggestion/tests/test_boa.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import numpy as np
+
+from ..bayesianoptimization.src.bayesian_optimization_algorithm import BOAlgorithm
+
+
+def test_boa(dim, request_num, lower_bounds, upper_bounds, X_train, y_train):
```
Comment (richardsliu) [inline pkg/suggestion/tests/test_boa.py]: General comment (for all test files): 
For all the tests, add a brief comment section that explains what this test does, what are the expectations, etc. This makes it easier for other contributors to follow.
